src/transport/websocket.ts

```diff
@@ -0,0 +1,444 @@
+    this.subscriptions = {};
+  }
+
+  public addNew (
+    path: string,
+    listeners: SubscriptionListeners,
+    headers: ElementsHeaders
+  ): number {
+    this.lastID = this.lastID + 1;
+
+    return this.add(
+      this.lastID,
+      path,
+      listeners,
+      headers
+    );
+  }
+  
+  public add (
+    subID: number,
+    path: string,
+    listeners: SubscriptionListeners,
+    headers: ElementsHeaders
+  ): number {
+    this.subscriptions[subID] = {
+      path,
+      listeners,
+      headers
+    };
+
+    return subID;
+  }
+
+  public has (subID: number): boolean {
+    return this.subscriptions[subID] != undefined;
+  }
+
+  public remove (subID: number): boolean {
+    return delete this.subscriptions[subID];
+  }
+
+  public get (subID: number): SubscriptionData {
+    return this.subscriptions[subID];
+  }
+
+  public getAll (): WsSubscriptionsType {
+    return this.subscriptions;
+  }
+
+  public getAllAsArray (): SubscriptionData[] {
+    return Object.keys(this.subscriptions).map(subID => (
+      {subID: parseInt(subID), ...this.subscriptions[subID]}
+    ));
+  }
+
+  public removeAll () {
+    this.subscriptions = {};
+  }
+
+}
+
+class WsSubscription implements Subscription {
+  constructor(
+    private wsTransport: WebSocketTransport,
+    private subID: number
+  ) {
+
+  }
+
+  unsubscribe () {
+    this.wsTransport.unsubscribe(this.subID);
+  }
+
+}
+
+const pingIntervalMs: number = 2000;
+const pingTimeoutMs: number = 2000;
+
+export default class WebSocketTransport implements SubscriptionTransport {
+  private baseURL: string;
+  private webSocketPath: string = '/ws';
+  private socket: WebSocket;
+  private subscriptions: WsSubscriptions;
+  private pendingSubscriptions: WsSubscriptions;
+  private lastMessageReceivedTimestamp: number;
+  private pingInterval: any;
+  private pongTimeout: any;
+  private lastSentPingID: number;
+
+  constructor(host: string) {
+    this.baseURL = `wss://${host}${this.webSocketPath}`;
+    this.subscriptions = new WsSubscriptions();
+    this.pendingSubscriptions = new WsSubscriptions();
+  
+    this.pingInterval = setInterval(() => {
+      if (this.pongTimeout) {
+        return;
+      }
+
+      const now = new Date().getTime();
+
+      if (pingTimeoutMs > (now - this.lastMessageReceivedTimestamp)) {
+        return;
+      }
+
+      this.sendMessage(
+        this.getMessage(
+          PingMessageType,
+          now
+        )
+      );
+
+      this.lastSentPingID = now;
+
+      this.pongTimeout = setTimeout(() => {
+        const now = new Date().getTime();
+
+        if (pingTimeoutMs > (now - this.lastMessageReceivedTimestamp)) {
+          this.pongTimeout = null;
+          return;
+        }
+
+        this.socket.close(4000, `Pong response wasn't recivied in timeout.`);
+      }, pingTimeoutMs);
+
+    }, pingIntervalMs);
+
+    this.connect();
+  }
+  
+  private connect () {
+    if (this.socket instanceof WebSocket) {
+      this.socket.close();
```

Comment:
If you call connect while a WebSocket connection is running and there are open subscriptions, this line will close the WebSocket, but won't clean up subscriptions. Those subscriptions will never receive events or even be closed.